docs/adr/transaction-log-state-store.md

```diff
@@ -0,0 +1,155 @@
+time.
+
+## Decision
+
+Use an event sourced model to store a transaction log, as well as snapshots of the state.
+
+Store the transactions as items in DynamoDB. Store snapshots as S3 files.
+
+The transaction log DynamoDB table has a hash key of the table ID, and range key of the transaction number in order. Use
+a conditional check to ensure the transaction number set has not been used.
+
+The snapshots DynamoDB table holds a reference to the latest snapshot held in S3, similar to the S3 state store
+revisions table. This also holds the transaction number that snapshot was derived from.
+
+## Consequences
+
+This should result in a similar update process to the S3 state store, but without the need to save or load the whole
+state at once. This should mean quicker updates even compared to the DynamoDB state store, since we only need to save
+one item per transaction. This would use a different set of patterns from those where the source of truth is a store of
+the current state, and we'll look at some of the implications.
+
+We'll look at how to model the state as derived from the transaction log, independent of the underlying store. To avoid
+reading every transaction every time, we can store a snapshot of the state, and start from a certain point in the log.
+
+We'll look at how to achieve ordering and durability of transactions. This is a slightly different approach for
+distributed updates, and there are potential problems in how we store the transaction log.
+
+We'll look at some applications for parallel models or storage formats, as this approach makes it easier to derive
+different formats for the state. This can allow for queries instead of loading the whole state at once, or we can model
+the data in some alternative ways for various purposes.
+
+We'll also look at how this compares to an approach based on a relational database.
+
+### Modelling state
+
+The simplest approach is to hold a model in memory for the whole state of a Sleeper table. We can use this one, local
+model for any updates or queries, and bring it up to date based on the ordered sequence of transactions. We just need to
+be able to apply any given transaction to the model.
+
+Whenever a change occurs, we create a transaction that we can apply to the model in memory. Anywhere that holds the
+model can bring itself up to date by reading only the transactions it hasn't seen yet, starting from the latest
+transaction that's already been applied locally. With DynamoDB, consistent reads can enforce that you're really
+up-to-date.
+
+We can also skip to a certain point in the transaction log. We have a separate process whose job is to write regular
+snapshots of the model. Every few minutes, we can take a snapshot against the latest transaction and write a copy of the
+whole model to S3. We can point to it in DynamoDB as in the S3 state store's revision table. This allows any process to
```

Comment:
Writing a copy of the tables state this often sounds like it would use a lot more storage in S3 than we would be using in the current implementation if we have particularly large tables. Perhaps we could have some kind of transaction manager that saves a copy of the table every N transactions?

Reply:
I think if the snapshots are less frequent than the updates, it would be a lot less than the S3 state store? We also don't need to hold old snapshots. We could add a way to clean them up.

Reply:
Right, makes sense. At the moment every update in the S3StateStore writes the state of the table to S3 so it would be less. Cleaning up the old models will definitely help reduce storage in S3.